README.md
# Vignette

Vignette is a Python lib for creating and fetching image thumbnails following the [Freedesktop specification](http://specifications.freedesktop.org/thumbnail-spec/thumbnail-spec-latest.html).

In a nutshell, the specification says that thumbnails should be 128x128 or 256x256 and should be cached in ~/.cache/thumbnails. If a thumbnail has already been generated for an URL, it can be fetched from the cache first instead of generating a new one.

# Documentation

The Python module is full of docstrings, but here's a short documentation.

This library has querying functions, that do not generate thumbnails, and can be used with files or URLs, that can be non-images:

* build_thumbnail_path
* try_get_thumbnail
* is_thumbnail_failed

It has functions that have side effects, which write thumbnails, or "fail-files" (if a thumbnail couldn't be generated), they can require local-files (see the function's doc):

* get_thumbnail
* create_thumbnail
* put_thumbnail
* put_fail

## Examples

Just ask for thumbnails of local images, automatically creating them if necessary:
```
  import vignette
  thumb_image = vignette.get_thumbnail('/my/file.jpg')
  local_app_display(thumb_image)
```

Ask for a thumbnail or generate it manually, for example a web-browser generating pages previews, that this module can't do itself:

```
  import vignette
  orig_url = 'http://example.com/file.pdf'
  thumb_image = vignette.try_get_thumbnail(orig_url, mtime=0) # mtime is not used in this example

  if not thumb_image:
    thumb_image = vignette.build_thumbnail_path(orig_url, 'large')
    try:
      local_app_make_preview(orig_url, thumb_image)
    except NetworkError:
      vignette.put_fail(orig_url, 'mybrowser-1.0', mtime=0)
    else:
      thumb_image = vignette.put_thumbnail(orig_url, 'large', mtime=0)
    if is_thumbnail_failed(orig_url):
      thumb_image = 'error.png'

  local_app_display(thumb_image)
```

# Requirements

Vignette works with both Python 2 and Python 3.

Vignette requires the Python Imaging Library (PIL) for now but can use other image backends as plugins.

# License

Vignette is licensed under the [WTFPLv2](http://wtfpl.net).

# Version

Vignette is currently at version 3.0.0 and uses [Semantic Versioning](http://semver.org/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [5.1.1] - 2023-09-05
### Added
- CI on gitlab to run tests

### Fixed
- require at least pillow >6

## [5.1.0] - 2023-04-28
### Added
- "vignette" logger
- pillow >10 support
- PyQt6 support

### Deprecated
- PyQt5 support

## [5.0.0] - 2021-09-11
### Added
- support new x-large (512px) and xx-large (1024px) sizes from standard
- cli: support "-n" arg to return a thumbnail only if it already exists
- thumbnail metadata now includes "Software" key with app name or vignette name

### Changed
- MagickBackend: for image files, add width/height info in thumbnail

### Removed
- dropped python2 support

### Fixed
- "TryExec" key can be missing in gnome-thumbnailer files
- PilBackend: fix bad original size saved in thumbnail metadata
- extra metadata passed to create_thumbnail is not ignored anymore
- for image source files, EXIF orientation is respected when creating thumbnail
- PilBackend: try to fully load thumbnail in case of attribute not found
- MagickBackend/QtBackend: don't expand images bigger than thumbnail size

## [4.5.2] - 2019-08-17
### Fixed
- add missing "handled_types" attributes and new FILETYPE_MISC type

## [4.5.1] - 2019-08-17
### Fixed
- add __version__ attribute because setup.cfg refered to it

## [4.5.0] - 2019-08-17
### Added
- backends: add oggThumb support for ogg videos
- backends: add exe-thumbnailer support for .exe icons
- backends: add evince-thumbnailer support for PDFs (and atril)
- detect appropriate thumbnail backends using libmagic
- and only call ThumbnailBackends that are appropriate for an input file
- group thumbnailers in categories so they can be enabled selectively

### Fixed
- QtBackend: get accepted mimes list from Qt itself
- QtBackend: test availability
- __main__ should return a non-zero value in case of failure

## [4.4.0] - 2018-05-07
### Added
- use GNOME thumbnailers from /usr/share/thumbnailers
- added a thumbnail dir linter tool

### Fixed
- put_thumbnail: don't set width/height in thumbnail metadata if missing

## [4.3.3] - 2017-01-21
### Fixed
- add VERSION.txt to source

## [4.3.2] - 2016-12-11
### Fixed
- fix ooo-thumbnailer backend seemingly returning an error
- do not ignore some MetadataBackend.update_backend errors

### Added
- add a __main__.py file for running with "-m vignette"

## [4.3.1] - 2016-12-03
### Fixed
- mention backends in module doc and package

## [4.3.0] - 2016-11-29
### Added
- add poppler backend
- add OOo backend
- add FFMpegBackend
- separate backend in 2 types: metadata and thumbnailing

### Fixed
- use unicode literals for closer Python 3 behavior
- backends get_info() method should return None in case of failure

## [4.2.1] - 2016-11-16
### Fixed
- don't fail if directories already exist
- convert floating point numbers MTime to int

### Security
- restrict thumbnail dirs permissions

## [4.2.0] - 2016-11-09
### Added
- add a PyQt backend for generating thumbnails
- more documentation

## [4.1.0] - 2016-10-13
### Added
- MagickBackend adds image dimensions in metadata
- more documentation


test.py
#!/usr/bin/env pytest

import hashlib
import os
from pathlib import Path
import shutil

import vignette
from pytest import fixture, skip


ALL_THUMBNAIL = vignette.THUMBNAILER_BACKENDS
IMAGE_THUMBNAIL = [vignette.QtBackend(), vignette.PilBackend(), vignette.MagickBackend()]
IMAGE_THUMBNAIL = [b for b in IMAGE_THUMBNAIL if b.is_available()]

ALL_METADATA = vignette.METADATA_BACKENDS
AVAIL_METADATA = [b for b in ALL_METADATA if b.is_available()]

SAMPLES_PATH = Path(__file__).parent / "samples"


@fixture()
def workdir(tmp_path):
	os.environ["XDG_CACHE_HOME"] = str(tmp_path)
	yield tmp_path


@fixture()
def image_src(workdir):
	shutil.copyfile(SAMPLES_PATH / "checkerboard.png", workdir / "test.png")
	src = str(workdir / "test.png")
	yield src


def get_backend_name(backend):
	return type(backend).__name__


@fixture(params=ALL_METADATA, ids=get_backend_name)
def metadata_backend(request):
	if not request.param.is_available():
		skip(f"{request.param} is not available")

	vignette.METADATA_BACKENDS = [request.param]
	vignette.THUMBNAILER_BACKENDS = [request.param]
	yield
	vignette.METADATA_BACKENDS = ALL_METADATA
	vignette.THUMBNAILER_BACKENDS = ALL_THUMBNAIL


def require_pil():
	try:
		import PIL.Image
	except ImportError:
		skip("the test requires PIL for checking correctness")


def get_metadata(path):
	import PIL.Image

	img = PIL.Image.open(path)
	# PIL might not load all metadata without load()
	# in particular, images saved with the magickbackend
	img.load()
	return img.info


def test_at_least_one_backend():
	assert any(backend.is_available() for backend in ALL_METADATA)


def test_hash(image_src, workdir, metadata_backend):
	uri = f"file://{image_src}"
	hash = hashlib.md5(uri.encode("utf-8")).hexdigest()
	dest = workdir / "thumbnails" / "large" / f"{hash}.png"

	assert str(dest) == vignette.build_thumbnail_path(image_src, "large")


def test_metadata(image_src, metadata_backend):
	require_pil()

	dest = vignette.create_thumbnail(image_src, "large")

	assert dest
	info = get_metadata(dest)
	assert info["Thumb::URI"] == f"file://{image_src}"
	assert str(int(os.path.getmtime(image_src))) == info["Thumb::MTime"]
	assert "512" == info["Thumb::Image::Width"]
	assert "512" == info["Thumb::Image::Height"]
	assert str(os.path.getsize(image_src)) == info["Thumb::Size"]


def test_dont_expand(workdir, metadata_backend):
	require_pil()
	import PIL.Image

	src_path = SAMPLES_PATH / "rose.jpg"

	shutil.copy(src_path, workdir)

	thumb_path = vignette.get_thumbnail(str(src_path))
	thumb_img = PIL.Image.open(thumb_path)

	assert thumb_img.size == (70, 46)


def test_shrink(image_src, metadata_backend):
	require_pil()
	import PIL.Image

	thumb_path = vignette.get_thumbnail(str(image_src))
	thumb_img = PIL.Image.open(thumb_path)
	assert thumb_img.size == (256, 256)

	thumb_path = vignette.get_thumbnail(str(image_src), size="normal")
	thumb_img = PIL.Image.open(thumb_path)
	assert thumb_img.size == (128, 128)

	thumb_path = vignette.get_thumbnail(str(image_src), size="x-large")
	thumb_img = PIL.Image.open(thumb_path)
	assert thumb_img.size == (512, 512)

	thumb_path = vignette.get_thumbnail(str(image_src), size="xx-large")
	thumb_img = PIL.Image.open(thumb_path)
	assert thumb_img.size == (512, 512)


def test_basic(image_src, metadata_backend):
	dest = vignette.build_thumbnail_path(image_src, "large")
	assert dest

	assert vignette.try_get_thumbnail(image_src, "large") is None
	assert not os.path.exists(dest)

	assert dest == vignette.get_thumbnail(image_src, "large")
	assert os.path.isfile(dest)

	assert dest == vignette.try_get_thumbnail(image_src, "large")


def test_reuse_thumbnail(image_src, metadata_backend):
	dest = vignette.get_thumbnail(image_src, "large")
	assert dest

	st = os.stat(dest)
	assert dest == vignette.get_thumbnail(image_src, "large")
	assert st == os.stat(dest)
	assert dest == vignette.create_thumbnail(image_src, "large")
	assert st != os.stat(dest)


def test_direct_thumbnail(image_src, metadata_backend):
	dest = vignette.get_thumbnail(image_src, "large")
	assert dest
	assert os.path.isfile(dest)
	assert dest == vignette.try_get_thumbnail(dest, "large")


def test_mtime_validity(image_src, metadata_backend):
	dest = vignette.get_thumbnail(image_src, "large")
	assert dest

	os.utime(image_src, (0, 0))
	assert vignette.try_get_thumbnail(image_src, "large") is None
	assert os.path.isfile(dest)

	assert dest == vignette.get_thumbnail(image_src, "large")
	assert dest == vignette.try_get_thumbnail(image_src, "large")


def test_multisize(image_src, metadata_backend):
	dest = vignette.get_thumbnail(image_src, "large")
	assert dest
	assert dest == vignette.try_get_thumbnail(image_src, "large")
	assert vignette.try_get_thumbnail(image_src, "normal") is None
	assert dest == vignette.try_get_thumbnail(image_src)

	os.remove(dest)
	dest = vignette.get_thumbnail(image_src, "normal")
	assert dest
	assert dest == vignette.try_get_thumbnail(image_src, "normal")
	assert vignette.try_get_thumbnail(image_src, "large") is None
	assert dest == vignette.try_get_thumbnail(image_src)


def test_fail(workdir, metadata_backend):
	src = workdir / "empty"
	src.touch()
	src = str(src)

	assert vignette.get_thumbnail(src, "large") is None
	assert not (workdir / "thumbnails" / "fail").exists()
	assert not vignette.is_thumbnail_failed(src, "foo")

	assert vignette.get_thumbnail(src, "large", use_fail_appname="foo") is None
	assert (workdir / "thumbnails" / "fail" / "foo").exists()
	assert vignette.is_thumbnail_failed(src, "foo")
	assert not vignette.is_thumbnail_failed(src, "bar")


def test_fail_mtime_validity(workdir, metadata_backend):
	src = workdir / "empty"
	src.touch()
	src = str(src)

	assert vignette.get_thumbnail(src, "large", use_fail_appname="foo") is None
	assert vignette.is_thumbnail_failed(src, "foo")

	os.utime(src, (0, 0))
	assert not vignette.is_thumbnail_failed(src, "foo")


def test_put_thumbnail(image_src, metadata_backend):
	uri = "https://example.com"
	tmp = vignette.create_temp("large")
	shutil.copyfile(image_src, tmp)
	vignette.put_thumbnail(uri, "large", tmp, mtime=42)
	assert vignette.try_get_thumbnail(uri, "large", mtime=42)
	assert vignette.try_get_thumbnail(uri, "large", mtime=1) is None


def test_put_fail(image_src, metadata_backend):
	# force a "failure"
	vignette.put_fail(image_src, "foo")

	assert vignette.is_thumbnail_failed(image_src, "foo")
	assert not vignette.is_thumbnail_failed(image_src, "bar")
	# check the failure is found instead of creating a thumbnail
	assert vignette.get_thumbnail(image_src, use_fail_appname="foo") is None

	# create the thumbnail
	dest = vignette.get_thumbnail(image_src, use_fail_appname="bar")
	assert dest

	# check the failure is now ignored
	assert dest == vignette.get_thumbnail(image_src, use_fail_appname="foo")


def test_orientation(workdir, metadata_backend):
	require_pil()
	import PIL.Image

	correct_src = SAMPLES_PATH / "rose.jpg"
	oriented_src = SAMPLES_PATH / "rose-oriented.jpg"

	shutil.copy(correct_src, workdir)
	shutil.copy(oriented_src, workdir)

	correct_dst = vignette.get_thumbnail(str(correct_src))
	correct_img = PIL.Image.open(correct_dst)

	oriented_dst = vignette.get_thumbnail(str(oriented_src))
	oriented_img = PIL.Image.open(oriented_dst)

	assert correct_img.size == oriented_img.size


def test_moreinfo(image_src, metadata_backend):
	require_pil()

	moreinfo = {"foo": "bar", "Thumb::URI": "bad", "Thumb::MTime": "bad"}
	dest = vignette.create_thumbnail(image_src, "large", moreinfo=moreinfo)
	assert dest
	info = get_metadata(dest)

	assert info["foo"] == "bar"
	assert info["Thumb::URI"] != "bad"
	assert info["Thumb::MTime"] != "bad"


test_thumbnail.py


import sys, os
import commands
import unittest
import md5

import thumbnail as T


genRose = lambda x: os.system('convert rose: %s' % x)
exists = os.path.exists


class ThumbnailTests(unittest.TestCase):
	def md5(self, p):
		return md5.new(p).hexdigest()
	
	def path(self, m, size):
		return os.path.expanduser('~/.cache/thumbnails/%s/%s.png' % (size, m))
		
	def setUp(self):
		self.rp = '/tmp/rose.png'
		self.ru = 'file:///tmp/rose.png'
		self.rm = self.md5(self.ru)
		self.rt = os.path.expanduser('~/.cache/thumbnails/large/%s.png' % self.rm)
		genRose(self.rp)
		self.rmt = int(os.path.getmtime(self.rp))
		
		self.tp = '/tmp/test.txt'
		self.tu = 'file:///tmp/test.txt'
		self.tm = self.md5(self.tu)
		self.tt = os.path.expanduser('~/.cache/thumbnails/large/%s.png' % self.tm)
		self.tf = os.path.expanduser('~/.cache/thumbnails/fail/test-1.0/%s.png' % self.tm)
		os.system(': > /tmp/test.txt')
		self.tmt = int(os.path.getmtime(self.tp))
		
		self.rmRose()
		self.rmTxt()
		
		
	def rmRose(self):
		try:
			os.unlink(self.rt)
		except Exception:
			pass
	
	def rmTxt(self):
		try:
			os.unlink(self.tf)
		except Exception:
			pass
	
	def testNotExisting(self):
		self.assertFalse(T.is_thumbnail_failed(self.rp, 'test-1.0'))
		self.assertFalse(T.is_thumbnail_failed(self.ru, 'test-1.0'))
		
		self.assertEqual(T.thumbnail_path(self.rp, 'large'), self.rt)
		self.assertEqual(T.thumbnail_path(self.ru, 'large'), self.rt)
		self.assertEqual(T.thumbnail_path(self.rp, 256), self.rt)
		
		self.assertEqual(T.existing_thumbnail_path(self.rp, 'large'), False)
		self.assertEqual(T.existing_thumbnail_path(self.ru, 'large', self.rmt), False)
		self.assertEqual(T.existing_thumbnail_path(self.rp, 256), False)
		self.assertFalse(exists(self.rt))
		
	
	def testMFail(self):
		self.assertFalse(T.is_thumbnail_failed(self.tp, 'test-1.0'))
		self.assertFalse(T.is_thumbnail_failed(self.tu, 'test-1.0'))
		self.assertFalse(exists(self.tf))
		
		T.put_fail(self.tp, 'test-1.0')
		
		self.assert_(T.is_thumbnail_failed(self.tp, 'test-1.0'))
		self.assert_(T.is_thumbnail_failed(self.tu, 'test-1.0'))
		self.assert_(exists(self.tf))
		
		self.rmTxt()
	
	def testAFail(self):
		self.assertFalse(T.is_thumbnail_failed(self.tp, 'test-1.0'))
		self.assertFalse(exists(self.tf))

		T.gen_image_thumbnail(self.tp)
		self.assertFalse(T.is_thumbnail_failed(self.tp, 'test-1.0'))
		self.assertFalse(exists(self.tf))
		
		T.gen_image_thumbnail(self.tp, use_fail_appname='test-1.0')
		self.assert_(T.is_thumbnail_failed(self.tp, 'test-1.0'))
		self.assert_(exists(self.tf))
		
		self.rmTxt()
	
	def testThumb(self):
		self.assertFalse(T.existing_thumbnail_path(self.rp))
		self.assertFalse(exists(self.rt))
		
		T.gen_image_thumbnail(self.rp)
		self.assert_(T.existing_thumbnail_path(self.ru, mtime=self.rmt))
		self.assert_(exists(self.rt))
		
		self.rmRose()
		
		self.assertFalse(T.existing_thumbnail_path(self.ru, mtime=self.rmt))
		self.assertFalse(exists(self.rt))
		
		T.gen_image_thumbnail(self.rp)
		self.assert_(T.existing_thumbnail_path(self.rp))
		self.assert_(exists(self.rt))
		
	def testMod(self):
		T.gen_image_thumbnail(self.rp)
		self.assert_(T.existing_thumbnail_path(self.rp))
		self.assert_(exists(self.rt))
		
		os.system('touch -r / %s' % self.rp)
		self.assertFalse(T.existing_thumbnail_path(self.rp))
		self.assert_(exists(self.rt))
	
	def tearDown(self):
		self.rmRose()
		self.rmTxt()
		
		try:
			os.unlink(self.rp)
		except Exception:
			pass
			
		try:
			os.unlink(self.tp)
		except Exception:
			pass
		


unittest.main()


tools/thumbnails_lint.py
#!/usr/bin/env python3

"""Tool to clean obsolete thumbnails in ~/.cache/thumbnails
"""

import re
import os
from urllib.request import url2pathname
from urllib.parse import urlparse

import vignette


DELETE_EXTRA = True
COUNT = 0
SIZE = 0

def remove(t):
	global COUNT, SIZE

	COUNT += 1
	SIZE += os.path.getsize(t)
	os.unlink(t)


def do_dir(d):
	global COUNT

	fn_template = re.compile('%s/[0-9a-fA-F]{32}.png$' % re.escape(d))

	for f in os.listdir(d):
		f = os.path.join(d, f)

		if not fn_template.match(f):
			if DELETE_EXTRA:
				print('Extra file %r' % f)
				remove(f)
			continue

		info = vignette.thumbnail_info(f)
		if not info:
			print('Error parsing thumbnail %r' % f)
			remove(f)
			continue

		if not info.get('uri'):
			print('Invalid URI in thumbnail %r' % f)
			remove(f)
			continue

		uri_info = urlparse(info['uri'])
		if uri_info.scheme != 'file':
			continue

		target = url2pathname(uri_info.path)

		if not os.path.isfile(target):
			print('Missing file %r' % target)
			remove(f)
			continue

		if int(os.path.getmtime(target)) != info['mtime']:
			print('Different mtime of %r' % target)
			remove(f)
			continue


if __name__ == '__main__':
	for d in ('large', 'normal'):
		d = os.path.join(vignette._thumb_path_prefix(), d)
		if os.path.isdir(d):
			do_dir(d)
	print('Removed %d files (%d bytes)' % (COUNT, SIZE))


vignette.py
#!/usr/bin/env python

# started 2009-03-10
# 2009-03-14
# license: WTFPLv2

# TODO handle more exceptions
# TODO support thumbnails smaller than 128x128
# TODO .thumblocal
# TODO allow mtime no-check


"""Generate and retrieve thumbnails according to the `Freedesktop.org thumbnail standard`_.

Summary of the thumbnail standard
=================================
* Thumbnails of any file or URL (should it be an image, a video, a webpage) can be stored in ~/.cache/thumbnails in PNG format
* Two sizes are used for thumbnails : 128x128 and 256x256, stored respectively in ~/.cache/thumbnails/normal and ~/.cache/thumbnails/large
* files or URLs thumbnailed must have a modification time (mtime for short) to detect obsolescence of thumbnails
* additional metadata can be put in thumbnails, as key/value pairs, in the PNG text fields, such as the time length for a video file
* there are two attributes required to be put in the metadata : the source URI and the mtime
* if a thumbnail can't be generated by an app (e.g. because of an erroneus file), a "fail-file" can be written in ~/.cache/thumbnails/fail/appname-version

Module functions
================
The module's functions take care of putting the two mandatory attributes in the thumbnail file.

Functions for querying, that do not generate thumbnails, and can be used with files or URLs, that can be non-images:
* build_thumbnail_path
* try_get_thumbnail
* is_thumbnail_failed

Functions that have side effects, which write thumbnails, or "fail-files", they can require local-files (see the function's doc):
* get_thumbnail
* create_thumbnail
* put_thumbnail
* put_fail


Examples
========

Just ask for thumbnails of local images, automatically creating them if necessary::

  import vignette
  thumb_image = vignette.get_thumbnail('/my/file.jpg')
  local_app_display(thumb_image)

Ask for a thumbnail or generate it manually, for example a web-browser generating pages previews,
that this module can't do itself::

  import vignette
  orig_url = 'http://example.com/file.pdf'
  thumb_image = vignette.try_get_thumbnail(orig_url, mtime=0) # mtime is not used in this example

  if not thumb_image:
    thumb_image = vignette.build_thumbnail_path(orig_url, 'large')
    try:
      local_app_make_preview(orig_url, thumb_image)
    except NetworkError:
      vignette.put_fail(orig_url, 'mybrowser-1.0', mtime=0)
    else:
      thumb_image = vignette.put_thumbnail(orig_url, 'large', mtime=0)
    if vignette.is_thumbnail_failed(orig_url):
      thumb_image = 'error.png'

  local_app_display(thumb_image)


.. _Freedesktop.org thumbnail standard: https://specifications.freedesktop.org/thumbnail-spec/thumbnail-spec-latest.html

"""


import hashlib
import os
import re
import sys
import tempfile


__all__ = (
	'get_thumbnail',
	'try_get_thumbnail',
	'build_thumbnail_path',
	'create_thumbnail',
	'put_thumbnail',
	'put_fail',
	'is_thumbnail_failed',
	'KEY_WIDTH',
	'KEY_HEIGHT',
	'KEY_SIZE',
	'KEY_MIME',
	'KEY_DOC_PAGES',
	'KEY_MOVIE_LENGTH',
)

if sys.version_info.major == 2:
	bytes, str = str, unicode


KEY_URI = 'Thumb::URI'
KEY_MTIME = 'Thumb::MTime'
KEY_WIDTH = 'Thumb::Image::Width'
KEY_HEIGHT = 'Thumb::Image::Height'
KEY_SIZE = 'Thumb::Size'
KEY_MIME = 'Thumb::Mimetype'
KEY_DOC_PAGES =  'Thumb::Document::Pages'
KEY_MOVIE_LENGTH = 'Thumb::Movie::Length'


def _any2size(size):
	if size in ('large', 256, '256'):
		return (256, 'large')
	elif size in ('normal', 128, '128'):
		return (128, 'normal')
	else:
		try:
			if 0 < int(size) <= 128:
				return (128, 'normal')
		except ValueError:
			pass


_URI_RE = re.compile(r'[a-zA-Z0-9.+-]+:')


def _any2uri(sth):
	"""Get an URI from the parameter

	If it's already an URI, return it, else return a file:// URL of it
	"""

	if _URI_RE.match(sth):
		return sth
	else:
		return 'file://' + os.path.abspath(sth)


def _any2mtime(origname, mtime=None):
	if mtime is None:
		return int(os.path.getmtime(origname))
	else:
		return mtime


def _thumb_path_prefix():
	xdgcache = os.getenv('XDG_CACHE_HOME', os.path.expanduser('~/.cache'))
	return os.path.join(xdgcache, 'thumbnails')


def hash_name(src):
	uri = _any2uri(src)
	if isinstance(uri, str):
		uri = uri.encode('utf-8')

	return hashlib.md5(uri).hexdigest()


def is_thumbnail_failed(src, appname, mtime=None):
	"""Is the thumbnail for `name` failed with `appname`?

	:param src: the URL or path of the source file.
	:type src: str
	"""

	prefix = _thumb_path_prefix()
	apppath = os.path.join(prefix, 'fail', appname)
	uri = _any2uri(src)
	mtime = _any2mtime(src, mtime)
	md5uri = hash_name(src)
	thumb = os.path.join(apppath, '%s.png' % md5uri)
	return os.path.exists(thumb) and is_thumbnail_valid(thumb, uri, mtime)


def put_thumbnail(src, size, thumb=None, mtime=None, moreinfo=None):
	"""Put a thumbnail into the store.

	This method is typically used for thumbnailing non-image files (like PDFs, videos) or
	non-local files.

	The application creates the thumbnail image on its own, and pushes the thumbnail to the
	store with this function.

	:param src: the URL or path of the file thumbnailed.
	:type src: str
	:param size: desired size of thumbnail. Can be any of 'large', 256 for large
	             thumbnails or 'normal', 128 for small thumbnails.
	:param thumb: if None, path where the app created the thumbnail. The file will be copied
	              copied to the target.
	:param mtime: the modification time of the file thumbnailed. If `mtime` is None, `src`
	              has to be a local file and its mtime will be read.
	:param moreinfo: additional optional key/values to store in the thumbnail file.
	:type moreinfo: dict
	"""

	dest = build_thumbnail_path(src, size)

	if thumb is not None and dest != thumb:
		shutil.copyfile(thumb, dest)

	get_backend().update_metadata(src, dest, mtime, moreinfo)
	os.chmod(dest, 0o600)

	return dest


def put_fail(src, appname, mtime=None, moreinfo=None):
	"""Create a failed thumbnail info file.

	Creates directories if they don't exist.

	:param src: the URL or path of the file thumbnailed.
	:type src: str
	:param mtime: the modification time of the file thumbnailed. If `mtime` is None, `src`
	              has to be a local file and its mtime will be read.
	:param moreinfo: additional optional key/values to store in the thumbnail file.
	:type moreinfo: dict
	"""

	prefix = os.path.join(_thumb_path_prefix(), 'fail', appname)
	if not os.path.isdir(prefix):
		os.makedirs(prefix, 0o700)

	md5uri = hash_name(src)
	dest = os.path.join(prefix, '%s.png' % md5uri)
	get_backend().create_fail(src, dest, mtime, moreinfo)
	os.chmod(dest, 0o600)

	return dest


class PilBackend(object):
	@staticmethod
	def is_available():
		try:
			import PIL.Image
			import PIL.PngImagePlugin
		except ImportError:
			return False
		return True

	@staticmethod
	def _pnginfo(uri, mtime, moreinfo=None):
		import PIL.PngImagePlugin

		outinfo = PIL.PngImagePlugin.PngInfo()

		outinfo.add_text(KEY_URI, uri)
		outinfo.add_text(KEY_MTIME, str(mtime))

		if moreinfo:
			for k in moreinfo:
				outinfo.add_text(k, str(moreinfo[k]))

		return outinfo

	@classmethod
	def create_thumbnail(cls, src, dest, size, mtime, moreinfo=None):
		import PIL.Image

		try:
			img = PIL.Image.open(src)
		except IOError:
			return None

		uri = _any2uri(src)
		outinfo = cls._pnginfo(uri, mtime, moreinfo)
		outinfo.add_text(KEY_WIDTH, str(img.size[0]))
		outinfo.add_text(KEY_HEIGHT, str(img.size[1]))

		img.thumbnail((size, size), PIL.Image.ANTIALIAS)

		tmppath = tempfile.mkstemp(suffix='.png', dir=os.path.dirname(dest))
		os.close(tmppath[0])

		img.save(tmppath[1], pnginfo=outinfo)
		img.close()
		os.rename(tmppath[1], dest)
		return dest

	@classmethod
	def create_fail(cls, src, dest, size, mtime, moreinfo=None):
		import PIL.Image

		outinfo = cls._pnginfo(_any2uri(src), _any2mtime(src, mtime), moreinfo)

		img = PIL.Image.new('RGBA', (1, 1))
		img.save(dest, pnginfo=outinfo)
		img.close()

	@staticmethod
	def get_info(path):
		import PIL.Image

		img = PIL.Image.open(path)
		mtime = int(img.info[KEY_MTIME])

		res = {
			'mtime': mtime,
			'uri': img.info[KEY_URI],
		}
		img.close()
		return res

	@classmethod
	def update_metadata(cls, src, dest, mtime=None, moreinfo=None):
		img = PILI.open(dest)
		mtime = _any2mtime(src, mtime)
		outinfo = cls._pnginfo(_any2uri(src), mtime, moreinfo)
		img.save(dest, pnginfo=outinfo)
		img.close()


BACKENDS = [PilBackend()]

def get_backend():
	for backend in BACKENDS:
		if backend.is_available():
			return backend


def create_thumbnail(src, size, moreinfo=None, use_fail_appname=None):
	"""Generate a thumbnail for `filename`, even if the thumbnail existed.

	Returns the path of the thumbnail generated. Creates directories if they don't exist.

	If the thumbnail cannot be generated and `use_fail_appname` is given, a failure info file
	will be generated, associated to the given app name so it is not retried too often.

	:param src: path of the source file. Must be an image file. Cannot be a URL.
	:type src: str
	:param moreinfo: additional optional key/values to store in the thumbnail file.
	:type moreinfo: dict
	:param use_fail_appname: app name to use when creating a failure info.
	:type use_fail_appname: str
	"""

	size = _any2size(size)[0]
	dest = build_thumbnail_path(src, size)
	mtime = _any2mtime(src)
	filesize = os.path.getsize(src)

	if moreinfo is None:
		moreinfo = {}
	else:
		moreinfo = dict(moreinfo)
	moreinfo[KEY_SIZE] = str(filesize)

	if not os.path.isdir(os.path.dirname(dest)):
		os.makedirs(os.path.dirname(dest), 0o700)

	if get_backend().create_thumbnail(src, dest, size, mtime, moreinfo):
		os.chmod(dest, 0o600)
		return dest

	if use_fail_appname is not None:
		put_fail(src, use_fail_appname)


def build_thumbnail_path(src, size):
	"""Get the path of the potential thumbnail.

	The thumbnail file may or may not exist.

	:param src: path or URI of the source file.
	:type src: str
	:param size: desired size of thumbnail. Can be any of 'large', 256 for large
	             thumbnails or 'normal', 128 for small thumbnails.
	"""

	sizename = _any2size(size)[1]
	prefix = os.path.join(_thumb_path_prefix(), sizename)
	if src.startswith(prefix + '/'):
		return src

	md5uri = hash_name(src)
	return os.path.join(prefix, '%s.png' % md5uri)


def is_thumbnail_valid(thumbnail, uri, mtime):
	info = get_backend().get_info(thumbnail)
	return info['uri'] == uri and info['mtime'] == mtime


def try_get_thumbnail(src, size=None):
	"""Get the path of the thumbnail or None if it doesn't exist.

	:param src: path or URI of the source file.
	:type src: str
	:param size: desired size of thumbnail. Can be 'large', 256 or 'normal', 128. If None,
	             tries with the large thumbnail size first, then with the normal size.
	"""

	if size is None:
		sizes = ['large', 'normal']
	else:
		sizes = [size]

	mtime = _any2mtime(src)
	uri = _any2uri(src)

	for size in sizes:
		thumb = build_thumbnail_path(src, size)
		if os.path.exists(thumb):
			if src == thumb:
				return src # bypass checks, the URI won't match
			elif is_thumbnail_valid(thumb, uri, mtime):
				return thumb


def get_thumbnail(src, size=None, use_fail_appname=None):
	"""Get the path of the thumbnail and create it if necessary.

	Returns None if an error occured. Creates directories if they don't exist.

	If the thumbnail cannot be found, and a previous failure info file had been created with
	the given app name, the thumbnail generation is not attempted and None is returned.

	:param src: path of the source file. Must be an image file. Cannot be a URL.
	:type src: str
	:param size: desired size of thumbnail. Can be any of 'large', 256 for large
	             thumbnails or 'normal', 128 for small thumbnails. If None, searches for any size.
	:param moreinfo: additional optional key/values to store in the thumbnail file, if the thumbnail did not exist.
	:type moreinfo: dict
	:param use_fail_appname: app name to use when creating a failure info.
	:type use_fail_appname: str
	"""

	thumb = try_get_thumbnail(src, size)
	if thumb is not None:
		return thumb

	if use_fail_appname is not None:
		mtime = _any2mtime(src)
		if is_thumbnail_failed(src, use_fail_appname, mtime):
			return None

	if size is None:
		size = 'large'
	return create_thumbnail(src, size, use_fail_appname=use_fail_appname)


def main():
	print(get_thumbnail(sys.argv[1]))


if __name__ == '__main__':
	main()
